Fix error when adding two subjects, refs #12709

djjuhasz opened the pull request:
When submitting an information object that adds more than one new related resources (e.g. two subject access point terms) a 500 error would usually occur because one (or more) of the related resources (e.g. one of the subject terms) was not created successfully before trying to link it to the information_object, causing a foreign key constraint failure.

This fix involved multiple changes:
- Refactored the autocomplete.js code to reduce the amount of duplicate code, and clarify the flow,
- Changed how autcomplete.js submit logic works to ensure the primary form is submitted only after all iframes forms have been submitted returned a response
- Added logic to the QubitTransactionFilter to re-execute the code and SQL writes when MySQL returns a transaction deadlock error
- Removed some unnecessary and confusing calls to `QubitTransactionFilter::getConnection()`

djjuhasz committed:
Cosmetic: reformat autocomplete.js

Reformat with VS Code autoformatting

djjuhasz committed:
Fix multi-value autocomplete, refs #12709

When submitting an information object that adds more than one new related
resources (e.g. two subject access point terms) a 500 error would
usually occur because one (or more) of the related resources (e.g. one
of the subject terms) was not created successfully before trying to link
it to the information_object, causing a foreign key constraint failure.

This fix delays the submit of main (e.g. information_object) form until after
all new related resources have been completed (via iframe submit) and have
returned a response.

In addition I normalized the code for adding new rows to a multi-row
autocomplete, and for adding a new iframe to the DOM for creating new
related resources.

djjuhasz committed:
Don't use QubitTransactionFilter::getConnection()

The QubitTransactionFilter::getConnnection() is a Singleton wrapper around
Propel::getConnection() that doesn't include the optional arguments
supported by the latter method.

The name of QubitTransactionFilter::getConnection() can also give the mistaken
impression that the connection object returned by the method will have
different transaction behaviour than that returned by Propel::getConnection(),
which is not true (transactions are implemented in the symfony filter chain,
not in the connection object).

djjuhasz committed:
Retry execution on SQL deadlock, refs #12709

When MySQL throws a transaction deadlock error, rollback the PDO
transaction, then re-execute the whole filter stack, including the request
processing and database writes.  After 3 retries give up and throw an
exception.

I also moved the QubitTransactionFilter down the filter stack to execute
right before the cache and execution filters.

jraddaoui commented on apps/qubit/config/filters.yml:
```diff
@@ -25,5 +22,8 @@ QubitMeta:
 
+QubitTransaction:
+  class: QubitTransactionFilter
+
```
This filters can be extended/overwritten in the plugins, we should make sure those reference `QubitTransaction` now. I'm not sure if the reordering affects those extensions too but we should check that out too if the order is important.

djjuhasz replied:
Huh, okay I'll have a look at the plugins for filter config.

jraddaoui commented on lib/model/QubitJob.php:
```diff
@@ -275,7 +275,7 @@ public static function runJob($jobName, $jobParams = array())
       // processing the task. If we don't do this now the transaction will be committed
       // once this request is processed but not before the worker hits the database.
-      $connection = QubitTransactionFilter::getConnection();
+      $connection = Propel::getConnection();
```
I think both return the same connection, but in this case using `QubitTransaction` makes clearer from where the transaction comes. The comment may be good enough, but I'd leave this one like it was, just in case.

djjuhasz replied:
When I saw `QubitTransactionFilter::getConnection()` my assumption was that transaction is started by the `getConnection()` method call, whereas `Propel::getConnection()` doesn't start a transaction, which is false.  I think having a separate `QubitTransactionFilter::getConnection()` method that is basically just a wrapper for `Propel::getConnection()` is misleading and unnecessary.

I'm tempted to get rid of `QubitTransactionFilter::getConnection()` altogether, but I think it does help legibility and factoring in `QubitTransactionFilter::execute()`.  Maybe it's best to make `QubitTransactionFilter::getConnection()` a protected method to make it clear that it's not part of the public signature.

jraddaoui replied:
The transaction is started after getting the connection, when `$conn->beginTransaction();` is called. In the end `QubitTransactionFilter::getConnection()` calls `Propel::getConnection()` if it doesn't have a connection already. In both cases (I am almost sure) the same connection is obtained so I think it's not a big deal to use one or the other but, to be sure, calling `QubitTransactionFilter::getConnection()` gives me a bit more confident about having the same connection that started the transaction.

jraddaoui commented on lib/model/QubitObject.php:
```diff
@@ -150,23 +150,18 @@ public function save($connection = null)
 
   protected function insert($connection = null)
   {
-    if (!isset($connection))
-    {
-      $connection = QubitTransactionFilter::getConnection(QubitObject::DATABASE_NAME);
-    }
-
```
Why removing this one entirely? I'd do the same change that is done below.

djjuhasz replied:
`$connection` is only used in the local scope as an argument to `BaseObject::insert()` and `self::insertSlug()`, and both methods already use the same `$connection = Propel::getConnection()` pattern, so it's unnecessary here.

jraddaoui commented on lib/filter/QubitTransactionFilter.class.php:
```diff
@@ -32,6 +34,45 @@ public static function getConnection()
+      );
+    }
+
+    // Get the current action instance
+    $actionInstance = $this->context
+      ->getController()
+      ->getActionStack()
+      ->getLastEntry()
+      ->getActionInstance();
+
+    // Create a new filter chain and reload config
+    $filterChain = new sfFilterChain();
+    $filterChain->loadConfiguration($actionInstance);
+
+    $this->context->getEventDispatcher()->notify(
+      new sfEvent($this, 'transaction.retry')
+    );
+
+    // Execute whole filter chain again
+    $filterChain->execute();
```
:mage:

That's a nice one @djjuhasz! I didn't know you could something like this and I have a few questions ...

- It's rare to see actions chained together and I wonder if it even matters but, should we get the first action entry to fully retry the request?
- What would happen when the request cycle involves other processes apart from the DB transaction? I'm thinking in DO uploads for example, would the file be copied again, etc.?

jraddaoui replied:
Another minor thing, why the event dispatcher? I'd not add it unless it's listened somewhere else.

djjuhasz replied:
> It's rare to see actions chained together and I wonder if it even matters but, should we get the first action entry to fully retry the request?

My current theory is that the whole filter chain is executed for *every* action, and that only the current (last) action is relevant to $filter->execute(). This theory is based on:

1. The `$filterChain->loadConfiguration($actionInstance)` requires a single `$actionInstance`
2. The `sfController::forward()` method adds a new action to the action stack AND [creates and executes a new $filterChain](https://github.com/artefactual/atom/blob/qa/2.6.x/vendor/symfony/lib/controller/sfController.class.php#L224)
3. The `sfExecutionFilter()` only executes the [last action on the stack](https://github.com/artefactual/atom/blob/qa/2.6.x/vendor/symfony/lib/filter/sfExecutionFilter.class.php#L35) (it's where I got the code to get the $actionInstance in QubitTransactionFilter)

djjuhasz replied:
>What would happen when the request cycle involves other processes apart from the DB transaction? I'm thinking in DO uploads for example, would the file be copied again, etc.?

The request is processed before the filter chain is executed, but all of the code in the filter chain and action execution stages will be re-run.  I think it's necessary to run the filter/action code again when redoing the transaction to avoid problems like the nested set values being stale, and because there is no way (that I know of) with symfony or Propel to *just* replay all of the database read/writes in isolation from the action code execution.

I found this summary of the front controller execution from [Inside the Controller Layer](https://symfony.com/legacy/doc/gentle-introduction/1_4/en/06-Inside-the-Controller-Layer) "The Front Controller's Job in Detail" to be helpful:

1. Load the project configuration class and the symfony libraries.
2. Create an application configuration and a symfony context.
3. Load and initiate the core framework classes.
4. Load the configuration.
5. Decode the request URL to determine the action to execute and the request parameters.
6. If the action does not exist, redirect to the 404 error action.
7. Activate filters (for instance, if the request needs authentication).
8. Execute the filters, first pass.
9. Execute the action and render the view.
10. Execute the filters, second pass.
11. Output the response.

By re-executing the filter chain, I believe steps 7-10 are re-run, then the response is output (step 11).

djjuhasz replied:
> Another minor thing, why the event dispatcher? I'd not add it unless it's listened somewhere else.

Yes, fair enough.  I was copying the pattern from https://github.com/artefactual/atom/blob/qa/2.6.x/vendor/symfony/lib/controller/sfController.class.php#L224-L238 but in the absence of listeners, the event dispatcher notification is not doing anything.

jraddaoui replied:
> By re-executing the filter chain, I believe steps 7-10 are re-run, then the response is output (step 11).

Yeah, what worries me the most is the re-execution of step 9 but lets hope the actions that may cause a DB deadlock don't do a lot of not idempotent work outside the DB changes.

djjuhasz replied:
Yes, there's a lot of magic going on behind the curtain here.  I wish I knew a practical way to determine if this approach is creating regressions, but without good test coverage I can't see how. 🤷 As you say, hopefully the code execution is effectively idempotent outside of the DB transaction boundaries.

djjuhasz replied:
I guess worst-case scenario it won't take a lot of work to undo the transaction retry code, so if testing does turn up regressive behaviour, reversing the changes can be the exit strategy.

jraddaoui reviewed (changes requested):
Thanks a lot for looking into this @djjuhasz!

Quite some magic going on in that filter :smile: Please, see my comments and questions below, the only blocker would be the plugin filters.

PS. The actual changes from the autocomplete Javascript were quite hard to find. It's nice that you moved the cosmetic changes to a different commit, but I'd have appreciated another one for the reformatting part.

djjuhasz commented:
> PS. The actual changes from the autocomplete Javascript were quite hard to find. It's nice that you moved the cosmetic changes to a different commit, but I'd have appreciated another one for the reformatting part.

Sorry about that.  My development process for the autocomplete.js changes was not methodical - there was no "refactor" stage before changing the functionality, it was all mixed together.  I don't know any practical way to separate out a clean "refactor" commit after the fact in a case like this.

djjuhasz committed:
Fixes from code review feedback, refs #12709

- Remove QubitTransactionFilter::getConnection() method and call
Propel::getConnection() directly
- Remove event dispatcher notificatino for transaction retry, as there
are no listeners to be notified
- Change name and position of QubitTransaction filter in plugin
filter.yml files to match apps/qubit/config/filter.yml

jraddaoui commented on lib/filter/QubitTransactionFilter.class.php:
```diff
@@ -20,20 +20,9 @@
-
-    return self::$connection;
-  }
-
```
(╯°□°）╯︵ ┻━┻

jraddaoui reviewed (commented):
Thanks @djjuhasz, great work!

> Sorry about that. My development process for the autocomplete.js changes was not methodical - there was no "refactor" stage before changing the functionality, it was all mixed together. I don't know any practical way to separate out a clean "refactor" commit after the fact in a case like this.

Np David, maybe doing the reorganization into functions in a different commit would have helped, but not a big deal at this point.

djjuhasz commented:
Merged to qa/2.6.x: 0f22ac4520629bae5123a6c3953982ffa6b9fc77
